Add patches and files for supporting Axiado ax3000 chip

rogerli-ax opened the pull request:
The PR adds all patches and files for supporting Axiado ax3000 chip:

 1. Add patches for supporting Axiado AX3000 chip
                 Add supporting Axiado AX3000 chip and dcm3003 board.
                 Disable MMU and Data Cache for Axiado AX3000 chip.
                 Add supporting Axiado AX3000 dtb file.
                 Add using correct gpio.h file for Axiado AX3000 chip.
                 Add using hardcode device tree for Axiado AX3000_chip.
                 Add supporting Axiado MMC_SDHCI driver.
                 Add the function sdhci_set_uhs_timing() that is called
                 Make the function sdhci_set_clock() to be external so that can be called by axiado_sdhci.c.
                 Add supporting Axiado eMMC PHY driver
                 Add using Zynq serial driver for Axiado AX3000 chip.
                 Add disabling two configure flags for Axiado AX3000 chip.
                 Add power_domain_on() returning 0 if CONFIG_POWER_DOMAIN is disabled.
                 Add SDHCI controller UHS register offsets.
                 Add sdhci_set_clock() and sdhci_set_clock() function phototypes.

2. Add necessary Axiado specific files:
               arch/arm/dts/ax3000-evk.dts:
               arch/arm/dts/ax3000.dtsi:
                 Axiado AX3000 chip dts files.
               board/axiado/;
                 Axiado dcm3003 board supporting files.
               configs/ax3000_evk_defconfig
                 Axiado AX3000 configuration file.
               drivers/mmc/axiado_sdhci.c
                 Axiado SDHCI driver.
               drivers/phy/phy-axiado-emmc.c
                 Axiado EMMC Phy driver.
               include/configs/ax3000-evk.h:
                 Axiado AX3000 header file.

Faiz-Abbas committed:
mmc: sdhci: Make sdhci_set_clock() non static

The am654_sdhci driver needs to switch the clock off
before disabling its phy dll and needs to re-enable
the clock before enabling the phy again.

Therefore, make the sdhci_set_clock() function accessible
in the am654_sdhci driver.

Signed-off-by: Faiz Abbas <faiz_abbas@ti.com>
Reviewed-by: Tom Rini <trini@konsulko.com>

Faiz-Abbas committed:
mmc: sdhci: Add support for HOST_CONTROL2 and setting UHS timings

The HOST_CONTROL2 register is a part of SDHC v3.00 and not just specific
to arasan/zynq controllers. Add the same to sdhci.h.

Also create a common API to set UHS timings in HOST_CONTROL2.

Signed-off-by: Faiz Abbas <faiz_abbas@ti.com>
Reviewed-by: Tom Rini <trini@konsulko.com>

rajas-axiado committed:
CODEOWNERS: add codeowners file

twei-axiado commented on arch/arm/dts/ax3000-evk.dts:
```diff
@@ -0,0 +1,33 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (c) 2021-25 Axiado Corporation (or its affiliates). All rights reserved.
```
```suggestion
 * Copyright (c) 2021-25 Axiado Corporation (or its affiliates).
```

twei-axiado replied:
For GPL, I think we can't declare `All rights reserved.` Please revise other files too.

rogerli-ax replied:
Revised all Axiado files.

twei-axiado commented on arch/arm/dts/ax3000-evk.dts:
```diff
@@ -0,0 +1,33 @@
+
+/ {
+	model = "Axiado AX3000-EVK";
+	compatible = "axiado,ax3000-evk,", "axiado, ax3000";
```
```suggestion
	compatible = "axiado,ax3000-evk,", "axiado,ax3000";
```

rogerli-ax replied:
Done.

twei-axiado commented on arch/arm/dts/ax3000.dtsi:
```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-2.0+
+/*
+ * Copyright (c) 2021-25 Axiado Corporation (or its affiliates). All rights reserved.
```
```suggestion
 * Copyright (c) 2021-25 Axiado Corporation (or its affiliates).
```

rogerli-ax replied:
Done.

twei-axiado commented on arch/arm/include/asm/gpio.h:
```diff
@@ -1,5 +1,5 @@
 #if !defined(CONFIG_ARCH_UNIPHIER) && !defined(CONFIG_ARCH_STI) && \
-	!defined(CONFIG_ARCH_K3)
+	!defined(CONFIG_ARCH_K3) && !defined(CONFIG_TARGET_AX3000_EVK)
```
Why do we need this? Other macros are `CONFIG_ARCH_*`, I think adding `CONFIG_TARGET_*` here is not suitable.

rogerli-ax replied:
We need this due to compiling errors.

twei-axiado commented on arch/arm/Kconfig:
```diff
@@ -1142,6 +1142,19 @@ config TARGET_HIKEY
+	select MMC_SDHCI_AXIADO
+	select PHY_AXIADO_EMMC
+	help
+	  Support for Axiado AX3000 version1.0. AX3000 is a quad-core
```
```suggestion
	  Support for Axiado AX3000. AX3000 is a quad-core
```
I'm not sure if we should disclose the `Version 1.0`

rogerli-ax replied:
Done.

twei-axiado commented on board/axiado/dcm3003/dcm3003.c:
why `dcm`?

rogerli-ax replied:
-- hidden --

twei-axiado commented on board/axiado/dcm3003/Kconfig:
```diff
@@ -0,0 +1,22 @@
+# SPDX-License-Identifier: GPL-2.0+
+#
+#Copyright (c) 2021-25 Axiado Corporation (or its affiliates). All rights reserved.
```
```suggestion
# Copyright (c) 2021-25 Axiado Corporation (or its affiliates).
```

rogerli-ax replied:
Done.

twei-axiado commented on common/image.c:
```diff
@@ -1510,7 +1510,12 @@ int image_setup_linux(bootm_headers_t *images)
 
 	if (IMAGE_ENABLE_OF_LIBFDT && of_size) {
+/* for Axiado ax3000 skipping the memory setup of device tree as its already hardcoded */
+#if defined(CONFIG_TARGET_AX3000_EVK)
```
Adding a new macro like `IMAGE_SKIP_SETUP_LIBFDT` is more suitable than a `#ifdef`.

rogerli-ax replied:
Done.

twei-axiado commented on arch/arm/cpu/armv8/cache_v8.c:
```diff
@@ -467,8 +467,12 @@ void dcache_enable(void)
 	}
-
+/* Disable MMU and Data Cache for Axiado ax3000 */
+#if defined(CONFIG_TARGET_AX3000_EVK)
```
Why not just enable `CONFIG_SYS_DCACHE_OFF`?

rogerli-ax replied:
Enabled CONFIG_SYS_DCACHE_OFF.

rogerli-ax replied:
If  CONFIG_SYS_DCACHE_OFF is used, it will generate the compiling error:

aarch64-openbmc-linux-ld.bfd: common/built-in.o: in function `reserve_mmu':
| /User_data/home/smtong/new_work/u-boot-github/common/board_f.c:386:(.text.reserve_mmu+0x10): undefined reference to `get_page_table_size'
| Segmentation fault (core dumped)
| make[1]: *** [/User_data/home/smtong/new_work/u-boot-github/Makefile:1511: u-boot] Error 139
| make[1]: *** Deleting file 'u-boot'
| make: *** [Makefile:148: sub-make] Error 2
| make: Leaving directory '/User_data/home/smtong/new_work/u-boot-github'
| ERROR: oe_runmake failed
| WARNING: exit code 1 from a shell command.
NOTE: recipe u-boot-axiado-2022.01-r0: task do_compile: Failed

twei-axiado commented on drivers/serial/Kconfig:
```diff
@@ -723,7 +723,7 @@ config STM32_SERIAL
 config ZYNQ_SERIAL
 	bool "Cadence (Xilinx Zynq) UART support"
-	depends on DM_SERIAL && (MICROBLAZE || ARCH_ZYNQ || ARCH_ZYNQMP || ARCH_ZYNQMP_R5)
+	depends on DM_SERIAL && (MICROBLAZE || ARCH_ZYNQ || ARCH_ZYNQMP || ARCH_ZYNQMP_R5 || TARGET_AX3000_EVK)
```
We might need to add `ARCH_AX3000`. How do you think?

rogerli-ax replied:
We currently use TARGET_AX3000_EVK. 
I tried to add ARCH_AX3000, but I always got the error:

 In file included from /User_data/home/smtong/new_work/u-boot-KWS-7719/include/sdhci.h:14,
|                  from /User_data/home/smtong/new_work/u-boot-KWS-7719/drivers/mmc/sdhci.c:14:
| /User_data/home/smtong/new_work/u-boot-KWS-7719/arch/arm/include/asm/gpio.h:3:10: fatal error: asm/arch/gpio.h: No such file or directory
|     3 | #include <asm/arch/gpio.h>
|       |          ^~~~~~~~~~~~~~~~~

i  will do this later.

twei-axiado commented on include/command.h:
```diff
@@ -206,6 +206,11 @@ int board_run_command(const char *cmdline);
 #define CMD_FLAG_BOOTD		0x0002	/* command is from bootd	*/
 #define CMD_FLAG_ENV		0x0004	/* command is from the environment */
 
+#if defined(CONFIG_TARGET_AX3000_EVK)
```
We should not add target-related code here. You might consider undefining these settings in the target config

rogerli-ax replied:
Done.

twei-axiado commented on drivers/mmc/sdhci.c:
```diff
@@ -326,7 +326,7 @@ static int sdhci_execute_tuning(struct udevice *dev, uint opcode)
 }
 #endif
-static int sdhci_set_clock(struct mmc *mmc, unsigned int clock)
+int sdhci_set_clock(struct mmc *mmc, unsigned int clock)
```
These changes are getting from upstream. Can we use cherry-pick?
 3966c7d0060e
 d1c0a2200afb
 bda47bef7c5a

rogerli-ax replied:
Did git cherry-pick for 3966c7d and d1c0a22.

twei-axiado commented on include/power-domain.h:
```diff
@@ -144,9 +144,9 @@ static inline int power_domain_free(struct power_domain *power_domain)
 int power_domain_on(struct power_domain *power_domain);
 #else
-static inline int power_domain_on(struct power_domain *power_domain)
+static inline int power_domain_on(struct udevice *dev)
```
Why changes to udevice?

rogerli-ax replied:
Done.

rogerli-ax committed:
Add necessary patches and files for supporting Axiado ax3000

	1 Add patches for supporting Axiado AX3000 chip
	  arch/arm/Kconfig:
            Add supporting Axiado AX3000 chip and dcm3003 board.
	  arch/arm/cpu/armv8/cache_v8.c:
	    Disable MMU and Data Cache for Axiado AX3000 chip.
	  arch/arm/dts/Makefile:
  	    Add supporting Axiado AX3000 dtb file.
	  arch/arm/include/asm/gpio.h:
 	    Add using correct gpio.h file for Axiado AX3000 chip.
	  common/image.c:
            Add IMAGE_SKIP_SETUP_LIBFDT to skip LIBFDT setup
	  drivers/mmc/Kconfig:
  	    Add supporting Axiado MMC_SDHCI driver.
	  drivers/mmc/Makefile:
  	    Add compiling Axiado MMC SDHCI driver.
	  drivers/mmc/sdhci.c:
  	    Add fixing compiling errors.
	  drivers/phy/Kconfig:
  	    Add supporting Axiado eMMC PHY driver
	  drivers/phy/Makefile:
  	    Add compiling Axiado eMMC PHY driver.
  	  drivers/serial/Kconfig:
  	    Add using Zynq serial driver for Axiado AX3000 chip.
	  include/sdhci.h:
  	    Add renaming registers for fixing compiling errors.
	  include/power-domain.h:
  	    Add power_domain_on() returning 0 if CONFIG_POWER_DOMAIN
	    is disabled.
	  include/image.h:
  	    Add IMAGE_SKIP_SETUP_LIBFDT flag.

	2. Add necessary Axiado specific files:

       	  arch/arm/dts/ax3000-evk.dts:
            arch/arm/dts/ax3000.dtsi:
              Axiado ax3000 chip dts files.
            board/axiado/
              Axiado dcm3003 board supporting files.
            configs/ax3000_evk_defconfig
              Axiado ax3000 configuration file.
           drivers/mmc/axiado_sdhci.c
             Axiado SDHCI driver.
           drivers/phy/phy-axiado-emmc.c
             Axiado EMMC Phy driver.
           include/configs/ax3000-evk.h:
              Axiado ax3000 header file.

	Tested-by:     Siu Ming Tong <smtong@axiado.com>
        Signed-off-by: Siu Ming Tong <smtong@axiado.com>
